source/tutorials/2024-adding-bioinformatic-software-to-bioconda.rst

```diff
@@ -0,0 +1,460 @@
+
+You can see an example of a ``build.sh`` script below:
+
+.. code-block:: bash
+
+    #!/bin/bash
+
+    set -xe
+
+    export LDFLAGS="-L$PREFIX/lib"
+    export CPATH=${PREFIX}/include
+
+    mkdir -p $PREFIX/bin
+
+    case $(uname -m) in
+        aarch64)
+            CXXFLAGS="${CXXFLAGS} -fsigned-char"
+            ARCH_OPTS="SSE_FLAG= POPCNT_CAPABILITY=0"
+            ;;
+        *)
+            ARCH_OPTS=""
+            ;;
+    esac
+
+    make -j ${CPU_COUNT} CXX=$CXX RELEASE_FLAGS="$CXXFLAGS" ${ARCH_OPTS}
+    make install prefix=$PREFIX
+
+    cp evaluation/{centrifuge_evaluate.py,centrifuge_simulate_reads.py} $PREFIX/bin
+
+*A relatively simple example* `build.sh script for Centrifuge <https://github.com/bioconda/bioconda-recipes/blob/b95f209b980339300b2fd84514a4912f6ad495e9/recipes/centrifuge/build.sh>`_, *based on the descriptions above. Here it includes both `make install` compilation examples with Bioconda C++ environment variables and copying of the additional auxiliary scripts to the `bin/` directory.*
+
+However, as always, check other tools/packages for examples.
+
+Examples of small ``build.sh`` scripts from the four examples above:
+
+- `kallisto <https://github.com/bioconda/bioconda-recipes/blob/23fe8cc0729ff70883819a8d2b2fdfc4d1da1443/recipes/kallisto/build.sh>`_ (make install).
+- `MALT <https://github.com/bioconda/bioconda-recipes/blob/23fe8cc0729ff70883819a8d2b2fdfc4d1da1443/recipes/malt/build.sh>`_ (java jar file).
+- `metabinner <https://github.com/bioconda/bioconda-recipes/blob/23fe8cc0729ff70883819a8d2b2fdfc4d1da1443/recipes/metabinner/build.sh>`_ (auxiliary scripts).
+- `phynder <https://github.com/bioconda/bioconda-recipes/blob/23fe8cc0729ff70883819a8d2b2fdfc4d1da1443/recipes/phynder/build.sh>`_ (patching).
+- `grid <https://github.com/bioconda/bioconda-recipes/blob/23fe8cc0729ff70883819a8d2b2fdfc4d1da1443/recipes/grid/build.sh>`_ (database files).
+
+To provide further guidance based on my experience:
+
+The ``$PREFIX`` variable corresponds to the the root of the conda environment that eventually gets made on a users system when they install the conda package.
+You can explore our own conda environments to see what the ``$PREFIX`` looks like by running ``conda env list`` to see all of our own conda environments, and changing into the one of the directory listed in there.
+They often will look very similar to Unix root directories, with folders such as ``etc/``, ``bin/``, ``lib/``, ``share/``, etc.
+for example, if we have an executable or scripts that need to go into ``bin/``, we must copy this into ``$PREFIX/bin``.
+For some tools we may have to copy other files into other directories, such as databases [6]_, but this is less common.
+
+Another tricky thing is compiling of C++ code, which can be a bit of a pain.
+For reasons [7]_, we need to use specific variables that point to the non-standard (it seems) places that conda stores its libraries and headers.
```

Comment:
Not sure what those problems are? Can we just link to the c help?

In general it might be also good to link to the conda-forge documentation: https://conda-forge.org/docs/maintainer/adding_pkgs/#the-recipe-metayaml

Reply:
It seems to be one of the hardest type of compilation, based both on my experience and when I've looked for help in other PRs

The issues are often due to hardcoded varibles and paths, which conflict with bioconda's own compliler locations etc. It's also the one that I'm asked above more often at least.

The link to the C help is already in the next sentence :)

And for the conda-forge recipe yaml: I hadn't seen that before! it's very nice, I've added it to the top of the meta.yaml section.